### PR TITLE
Diamonds #4: implemented core operations for diamonds and splits - READY

### PR DESCRIPTION
**Pull request by fredbi**
~~Still very much WIP.
Those interested may take a look out of curiosity.~~
Feature complete, ready for review

NOTE to reviewers:  this PR gets on board the content of PRs Diamond #1, #2 and #3. 
You might want to focus here on the content of `pkg/core`, other locations being separated in the other PRs.

Signed-off-by: Frederic BIDON <frederic@oneconcern.com>

### CODE REVIEW

File: pkg/core/index.go [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+	}
+
+	chans.bundleEntries <- bundleEntriesRes{bundleEntries: entries, idx: index, id: id}
+}
+
+// downloadAll downloads all index files in parallel, given an iterator to point to the file location
+func (f *fileIndex) downloadAll(chans downloadBundleFileListChans) {
+	f.indexPather.Reset()
+
+	throttle := make(chan struct{}, f.concurrency)
```
Comment (ransomw1c) [inline pkg/core/index.go]: i think this is a semaphore?

i'm not clear what "throttle" means as a noun in terms of programming.  i usually think throttle is a verb?

not s'th to get too delve into too much either way.

Comment (fredbi) [reply pkg/core/index.go]: > A throttle is the mechanism by which fluid flow is managed by constriction or obstruction. An engine's power can be increased or decreased by the restriction of inlet gases, but usually decreased.

**Comment by ransomw1c** [comment]
reading this some.  my preliminary feedback can be backgrounded until it's out of draft mode.

one question is about the `patherIterator` api design, and there are at least two ways to phrase the question:

why does the indexer return a `func(int) string` instead of an iterator?  doesn't the `func` always get used by iterating from zero to the max value, the second element of the tuple returned by `patherIterator.Next`?

the other way to phrase it is, "what do we mean by indexing, exactly, and is the design conflating indexing as a concept with adjacent ideas like batching?"  like, when i think, "indexing," i'm usually thinking about ways to make database queries more efficient, and that could be what's going on here if we need different ways to slice and dice batches of files, using the same `func(int) string` over different sequences of integers.  if we're always using the same sequence of integers, why not encode that convention as described in the previous paragraph?

----

in the more general setting, this question is returning to the theme of how much to abstract with functions, and i get the sense that some subtle affinities about design decisions that aren't in agreement, where i typically want to abstract out a lot with first order functions and function closures but not quite as much as your designs typically do.

when i ask questions in on this theme, i'm mostly checking that my understanding is correct by asking if the alternate design _could_ suffice rather than suggesting that we _should_ follow the alternate design.

**Comment by ransomw1c** [comment]
still unreasonably preoccupied with the states for diamonds and splits, another quick idea is to represent the state for diamonds, for instance, as a list/slice with values among some enum/typed constants rather than a single enum/type constant value.

so for instance, rather than `DiamondCancelled`, we might have have `{DiamondInitialized, DiamondCancelling, DiamondDone}` and instead of `DiamondDone`, we'd have  {DiamondInitialized, DiamondCommitting, DiamondDone}`.

phrasing this another way, i'm pretty sure the transition matrix between the states as they currently exist is reasonably sparse:  there are some states that don't lead to other states, and it's a programming error if such a state transition occurs.

perhaps, if the states are encoded differently, we can prevent some class of errors, because the state transitions that these errors correspond to cannot possibly occur given the way state is encoded?

**Comment by fredbi** [comment]
@ransomw1c  about the iterator pattern - you rightly pointed out a difficult point.

- my main objective was to reuse code, not so much to reach perfect abstraction.
- "what do we mean by indexing": well this is not quite new, I am reusing the "bundle-files-{index}" stuff
-  code to download or upload the "files list file" is complex and ubiquitous. It was not possible to reuse directly the bundle code (I managed it for a bit, but these functions where needed in many places) 
- so i did it the other way: let's produce things that could be reused by bundles by refactoring bundle download/upload on our next spare cycle
- I am calling "files list file" an index file, mainly because "files list file" kinda stutters, and yes, I've a feeling that this is akin to a database index: you get your hash from there, otherwise, you get lost searching at random for your keys and hashes in the blob storage)

Eventually I came with this "iterator", which is more an "iterator giving an iterator" (but you now know that I wouldn't ever call things like "iteratorOfIterators"...).

It provides the ability to scan over all "bundle-files-{index}.yaml" files when iterated ONCE (typical examples: split upload, bundle upload part in diamond_commit)

In the most general case (collecting all files from all splits), it iterates over all splits which in turn iterate over all their "bundle-files-{index}.yaml" files.

**Comment by fredbi** [comment]
@ransomw1c the transitions indeed follow a very linear pattern.

I'd say that anything that would make this thing resemble a configurable automaton would be harder to read and understand. Anyhow, this is an interesting idea. If you have an example on how this pattern could be coded in a readable way, I am all in. But I am against the typical petri-net lib.


The state stuff is essentially here to protect against re-starting things unattended, or to commit before some splits are completed etc. Checks are implemented at the start of operations at the moment.

To be totally transparent, I am still struggling with some race conditions on the commit part, though... But we can deliver a _workable_ solution with still some small windows where there is a race condition.

**Comment by fredbi** [comment]
> why does the indexer return a func(int) string instead of an iterator or just another Next function that's a closure over an iterator? doesn't the func always get used by iterating from zero to the max value, the second element of the tuple returned by patherIterator.Next?

Mmh yes that is an interesting way of doing it. That would definitely clarify the many returned values. Adopting that. Thanks.

**Comment by ransomw1c** [comment]
even before reading the replies, i'd decided that the state representation stuff suggestion is potentially [cleverness](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=14m35s) [considered harmful](https://go-proverbs.github.io/), although perhaps not.  it could alternately be viewed as [choosing the correct data structures](http://users.ece.utexas.edu/~adnan/pike.html), and i don't have a clear idea what perspective is more accurate.

will avoid additional contemplation of proverbs for awhile.

File: pkg/core/index.go [hunk 2]
```diff
@@ -146,20 +106,22 @@ func (f *fileIndex) getFileIndex(pth string) (model.BundleEntries, error) {
...
+	entries, err := f.getIndexFile(pth)
 	if err != nil {
 		chans.error <- err
 		return
 	}
 
-	// TODO(fred): to be checked out but I don't think the index value in this output is used beyond logging
+	// TODO(fred): nice - diamond does not use the idx field, but bundle_unpack does,
+	// we keep it for this method to be reusable by bundle_unpack at a later time.
+	// Ideally, we shouldn't need that index numbering.
```
Comment (ransomw1c) [inline pkg/core/index.go]: the `idx` struct field was added in case it's necessary to preserve order between requests and results – i.e. whether we need index numbering in the affirmative case is a consequence of preserving sort order.

so i suggest determining whether preserving sort order is desirable, making decision about circulating indices through the channels based on that determination.

File: pkg/core/index.go [hunk 3]
```diff
@@ -11,50 +9,11 @@ import (
...
-	// This iterator may be called concurrently.
-	Next() (string, indexIterator)
-}
-
-// fileIndex builds an index of files from file lists in metadata.
+// fileIndex manages an index of files based on file lists stored in metadata.
+//
+// Indexing is critical to be able to retrieve files from the blob storage.
+//
+// File lists are split into manageable chunks. By default each single index file holds up to 1000 file entries.
```
Comment (ransomw1c) [inline pkg/core/index.go]: based on this comment, i now understand that "index" throughout this patch is a synonym (or near-synonym?) for "bundle filelist" aka `BundleEntries`.  mildly distressed about about having three synonyms, now, for this entity.  meanwhile, "context" could either mean the golang type that essentially amounts to immutable maps or the datamon type that's currently a collection of GCS stores.

i don't have such an issue with these naming conventions when reading the source, yet i worry that conversationally these naming conventions will eventually have a negative impact on the user experience.

this isn't specific to this patch, so i'm not saying this comment need block merge.  feel free to resolve after reading.

File: pkg/core/bundle_unpack.go [hunk 4]
```diff
@@ -155,6 +155,7 @@ func unpackBundleDescriptor(ctx context.Context, bundle *Bundle, publish bool) e
 type bundleEntriesRes struct {
 	bundleEntries model.BundleEntries
 	idx           uint64
+	id            string // the splitID when used in this context
```
Comment (ransomw1c) [inline pkg/core/bundle_unpack.go]: when we start talking about using a structure in multiple contexts, practically, what we have is a [sum type](https://github.com/fantasyland/daggy#daggytaggedsumtypename-constructors) type, also called an [algebraic data type](http://learnyouahaskell.com/making-our-own-types-and-typeclasses#algebraic-data-types), or [tagged union](https://en.wikipedia.org/wiki/Tagged_union), i guess is the wikipedia-official term.

i often find myself reaching for such a paradigm in golang only to recall that it doesn't provide tagged unions as a first-order concept, and [the standard reply](https://github.com/golang/go/issues/19412) about why not is that `interface`s are supposed to be used as functional replacements for tagged unions.

so i suggest doing just that:  create an interface type and use it instead of `bundleEntriesRes` when "context" in the sense of this comment cannot be determined from variables' types.  at the simplest, it could provide a [catamorphism](https://github.com/ransomw/kvsjs/blob/master/src/dsl.js#L26) (i.e. fan-out) from the `interface` type to the original `bundleEntriesRes` and an additional type similar to `bundleEntriesRes` except in contexts where `splitID` is used, although there are other interface designs available.

what i'm saying here, most generally, is, "let's try to encode structure of the source first, then use comments for clarification and generating prose-like documentation."

File: pkg/model/diamond.go [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+	return &DiamondDescriptor{
+		DiamondID: ksuid.String(),
+		StartTime: GetBundleTimeStamp(),
+		State:     DiamondInitialized,
+		Mode:      EnableConflicts,
+	}
+}
+
+// NewDiamondDescriptor builds a new DiamondDescriptor
+func NewDiamondDescriptor(opts ...DiamondDescriptorOption) *DiamondDescriptor {
```
Comment (ransomw1c) [inline pkg/model/diamond.go]: for consistency [:zombie:  ... err, ⚡️ , i guess?] , i suggest either placing this in `core` or moving `core.NewBDescriptor` to `model`.

consistency has a very early precedence for me by all indications.  i'm considering consistency more important that separation of ideas by patch here.

Comment (fredbi) [reply pkg/model/diamond.go]: yes this is a refactoring I have already prepared, but will push that PR _after_ diamonds.

Comment (ransomw1c) [reply pkg/model/diamond.go]: order of operations considered a nitpick, then.

File: pkg/core/diamond_commit.go [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+// Cancel a diamond
+func (d *Diamond) Cancel() error {
+	// TODO(fred)
+	return status.ErrNotImplemented.WrapMessage("diamond cancel")
+}
+
+// WithState returns the diamond with its state modified and transition timestamp updated
+func (d *Diamond) WithState(state model.DiamondState) *Diamond {
+	d.DiamondDescriptor.State = state
+	d.DiamondDescriptor.StateChangeTime = time.Now()
```
Comment (ransomw1c) [inline pkg/core/diamond_commit.go]: realizing i might be entering into the discussion too late on this implementation detail since i wasn't present for all of the design phase:

is there any way we can avoid storing times from the machines where the datamon binary runs and instead use times provided by the GCS service?  that way, we'll have the exceedingly nice property of synchronization.  [`ObjectAttrs`](https://godoc.org/cloud.google.com/go/storage#ObjectAttrs) keeps track of times.  i'd be happier with a patch that tears the `storage.Store` up a little (as we're already doing with conditionals CRC) in order to avoid unsynchronized timestamps appearing in the data sets that it's this program's responsibility to organize and track.

Comment (fredbi) [reply pkg/core/diamond_commit.go]: I understand the sync issue, but I don't know of any such service specifically for GCP. You might otherwise subscribe to time.google.com as your NTP server. I am guessing that's how our pods are subscribed to.

Comment (ransomw1c) [reply pkg/core/diamond_commit.go]: > subscribe to time.google.com as your NTP server. I am guessing that's how our pods

to iron out the user story somewhat, can we _either_ make sure

* pods are time-synced (@bluelion ?)

and

* users know (via the markdown in this readme, confluence, or just direct correspondence) that the `diamond` and `split` functionality is mainly intended for in-k8s programs, and we should be notified in #datamon or otherwise if `diamond` and `split` are used from local machines, at least for data sets that might find their way into production

_or_

* include a ± time constant to encode the uncertainty, used along with (or to derive) `diamondCommitLockExpiry` and any other future compairsions with `StateChangeTime`?

File: pkg/core/diamond_commit.go [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+	if d.BundleID == "" {
+		return nil
+	}
+
+	// case of bundleID preservation (e.g. data migration use case)
+	id, err := ksuid.Parse(d.BundleID)
+	if err != nil {
+		return status.ErrInvalidKsuid.Wrap(err)
+	}
+	d.setBundleID(id.String())
```
Comment (ransomw1c) [inline pkg/core/diamond_commit.go]: specifically, this line could be removed, except for the data migration?

is it worth feature-flagging this use case?

File: pkg/core/diamond_commit.go [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+
+	// check when bundle is already specified (e.g. data migration use case)
+	if err = d.checkBundleID(); err != nil {
+		return err
+	}
+
+	logger := d.l.With(zap.String("diamond_id", d.DiamondDescriptor.DiamondID))
+
+	// critical section #1
+	// TODO(fred): better protect critical section
```
Comment (ransomw1c) [inline pkg/core/diamond_commit.go]: recording consideration of state-management metaphors for GCS while here for future reference:

do lockfiles translate to gcs?

"mutations" on immutable objects are implemented by creating a new object, and immutability has some desirable properties.  so state could be tracked via a directory-like prefix `.../<diamond-prefix>/<state-name>/` and then use some lexicographical ordering (e.g. zero-padded integers) on objects to distinguish the current state from previous states.

Comment (fredbi) [reply pkg/core/diamond_commit.go]: yes, using such resources as "lock-files" was my initial design.
However, since we can't delete the files, and since state is not always going forward but may revert (from running to initialized, from failed to running, etc), I gave up that way.
The downside of this is now I have a race condition on updating gcs files.

File: pkg/core/diamond_commit.go (hunk 6 above)
Comment (fredbi) [reply pkg/core/diamond_commit.go]: indeed you have a point here - I don't have a good solution for it right now. Mentioned it in the detailed design document. See https://github.com/oneconcern/datamon/pull/399

File: pkg/core/diamond_commit.go (hunk 7 above)
Comment (fredbi) [reply pkg/core/diamond_commit.go]: added feature guard

File: pkg/core/bundle_unpack.go (hunk 4 above)
Comment (fredbi) [reply pkg/core/bundle_unpack.go]: With this PR, I am trying to reuse what I can from existing bundle stuff.
A lot of such interesting theoretical points indeed come up to mind when doing so.
The practical outcome is what I did. The comment stuff is left behind so I may refactor this later on.

File: pkg/core/index.go (hunk 3 above)
Comment (fredbi) [reply pkg/core/index.go]: index stands for "file list file"

File: pkg/core/index.go (hunk 2 above)
Comment (fredbi) [reply pkg/core/index.go]: This investigation shall be conducted. From what I understand from the existing bundle functionality, this order preserving idea doesn't mean much, but I'll make sure later on.

Indeed, I recently drafted some experimental refactoring of this bundle download/upload stuff, which comes clearer now that I have to implement a more general functionality with 90% overlap. This draft is not published yet.

File: docs/proposals/diamond-design.md [hunk 9]
```diff
@@ -0,0 +1,562 @@
+# Diamond workflow support
+
+A diamond workflow allows for a pipeline to scale out the generation of a single bundle.
```
Comment (kerneltime) [inline docs/proposals/diamond-design.md]: Should we close https://github.com/oneconcern/datamon/pull/399

File: docs/proposals/diamond-design.md [hunk 10]
```diff
@@ -0,0 +1,562 @@
...
+                      [--with-conflicts]   # <- this is the default - overlaps are reported as conflicts detected when merging splits
+
+datamon diamond commit --diamond {diamond-id} \
+                       --with-checkpoints   # <- indicates an incremental upload: conflicts are subsequent versions of uploaded files
+
+datamon diamond commit --diamond {diamond-id} \
+                       --no-conflicts       # <- forbids any overlap to take place at commit time
+
+datamon diamond commit --diamond {diamond-id} \
+                       --ignore-conflicts   # <- do not report about any conflict: intermediate versions
```
Comment (kerneltime) [inline docs/proposals/diamond-design.md]: This is scope creep, this is also unsafe as a client can lose data and we will allow it.

File: docs/proposals/diamond-design.md [hunk 11]
```diff
@@ -0,0 +1,562 @@
...
+with conflicts. See also [§ Handling Conflicts below](#handling-conflicts).
+
+
+Available options at commit time:
+```
+datamon diamond commit --diamond {diamond-id} \
+                      [--with-conflicts]   # <- this is the default - overlaps are reported as conflicts detected when merging splits
+
+datamon diamond commit --diamond {diamond-id} \
+                       --with-checkpoints   # <- indicates an incremental upload: conflicts are subsequent versions of uploaded files
```
Comment (kerneltime) [inline docs/proposals/diamond-design.md]: I still prefer the client reusing the same split-id and letting the most recent file win. I don't yet see the reason to have this flag.

File: docs/proposals/diamond-design.md [hunk 12]
```diff
@@ -0,0 +1,562 @@
...
+   This is a consequence of the blob deduplication strategy (the same blob might be claimed by another file, so we can't
+   mark it as something to be deleted).
+
+
+Design:
+
+* Diamonds abide by the following state-transition model:
+  1. `initialized`
+  2. `committing`
+  3. `done`
```
Comment (kerneltime) [inline docs/proposals/diamond-design.md]: Instead of tracking this as fields in a single yaml file which will require a Read-Modify-Write cycle which under concurrent load is hard to get right. Use separate files and use the put if not present primitive. Also we only need init and done.

Comment (kerneltime) [reply docs/proposals/diamond-design.md]: I don't see the need for a 2 phase commit verbiage.

File: docs/proposals/diamond-design.md [hunk 13]
```diff
@@ -0,0 +1,562 @@
...
+
+2. Concurrent splits.
+  * Each `split add` command runs with its uniquely generated internal ID: metadata can be safely modified within this namespace.
+
+3. Commit consistency.
+  * The diamond is in effect concluded with the first commit associated with it. No further action is allowed on a committed diamond.
+  * Any attempt to commit while another commit is in progress is rejected (the operation should normally be fast
+    since only metadata is affected).
+  * For the duration of the commit operation, the diamond is in effect locked in "committing" status.
+  * To recover from uncontrolled failures during a commit (e.g. pod crash,...) this lock is considered stale after some
```
Comment (kerneltime) [inline docs/proposals/diamond-design.md]: This is not needed, we only need a single done.yaml file in the path with the bundle ID in it to mark commit. The put if not present primitive of cloud storage APIs will provide the necessary protection.

**Comment by kerneltime** [comment]
Not completed the review, have some concerns about state tracking which seems redundant and more than what is needed without serving a purpose.

File: docs/proposals/diamond-design.md (hunk 9 above)
Comment (fredbi) [reply docs/proposals/diamond-design.md]: I made (and sync'ed) all these PRs to produce a _reviewable_ ensemble.
I'd prefer that we merge them one by one following informed and understood agreement.

File: docs/proposals/diamond-design.md (hunk 10 above)
Comment (fredbi) [reply docs/proposals/diamond-design.md]: I can disable this if you prefer.

File: docs/proposals/diamond-design.md (hunk 12 above)
Comment (fredbi) [reply docs/proposals/diamond-design.md]: _verbiage_ is perhaps not carrying the same meaning. Certainly not when compared with some _actual_ verbiage that clutters the other reviews on this series of PR.
Best to discuss that part separately.

Comment (fredbi) [reply docs/proposals/diamond-design.md]: moved back to fewer states for diamond: initialized (aka running), done, canceled

and for splits: running, done

Comment (fredbi) [reply docs/proposals/diamond-design.md]: Even though the key name now holds the state semantics, I am keeping the state encoded in the metadata for easier retrieval (it is a bit of a pain to query the state of things when all is encoded as keys)

File: docs/proposals/diamond-design.md (hunk 13 above)
Comment (fredbi) [reply docs/proposals/diamond-design.md]: All that part about locking is indeed no more needed since the file no more needs to be locked.

File: docs/proposals/diamond-design.md (hunk 10 above)
Comment (fredbi) [reply docs/proposals/diamond-design.md]: --ignore-conflict is not unsafe: the result is the same with all flags. The only difference lies in the presence of /.conflicts, /.checkpoints "hidden" folders.

File: pkg/core/split.go [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+
+	if s.SplitDescriptor.Tag != "" {
+		s.l = s.l.With(zap.String("tag", s.SplitDescriptor.Tag))
+		s.Bundle.l = s.l
+	}
+	return s
+}
+
+// Upload a dataset as a split, without committing as a bundle
+func (s *Split) Upload(opts ...Option) error {
```
[the remark this answers is not in the document]
Comment (fredbi) [reply pkg/core/split.go]: yes, it is part of the interface: it is called by the CLI (next PR)

File: pkg/core/split.go [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+	"github.com/oneconcern/datamon/pkg/model"
+	"github.com/oneconcern/datamon/pkg/storage"
+	"go.uber.org/zap"
+	"gopkg.in/yaml.v2"
+)
+
+// Split is a generalization of a Bundle, to support diamond operations
+type Split struct {
+	metaObject
+	*Bundle
```
[the remark this answers is not in the document]
Comment (fredbi) [reply pkg/core/split.go]: this is type composition: it is quite different from `metaObject *Bundle`
Basically this means that a split is composed of a `metaObject` and a `*Bundle`
which also means a split inherits from all their methods.

File: pkg/core/split.go [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+	uploadIndexer   *fileIndex
+	downloadIndexer *fileIndex
+	mustExist       bool // used to check replayed splits with splitID forced
+}
+
+func defaultSplit(repo, diamondID string, stores context2.Stores) *Split {
+	empty := NewBundle(NewBDescriptor(), Repo(repo), ContextStores(stores))
+	split := model.NewSplitDescriptor()
+	return &Split{
+		metaObject:      defaultMetaObject(stores.VMetadata()), // splits live on vmetadata
```
[the remark this answers is not in the document]
Comment (fredbi) [reply pkg/core/split.go]: yep sure. My bad here.

File: pkg/core/diamond_commit.go [hunk 17]
```diff
@@ -0,0 +1,418 @@
...
+					continue
+				}
+
+				existing := obj.(mergeEntry)
+				if file.Hash == existing.Hash {
+					continue
+				}
+
+				if file.SnapshotTime.IsZero() {
+					panic("dev error: files should have a timing") // internal safeguard
```
[the remark this answers is not in the document]
Comment (fredbi) [reply pkg/core/diamond_commit.go]: yep sure
